### PR TITLE
Update Spring APIs to Spring Boot 3

### PR DESCRIPTION
**Pull request by aureamunoz**
### Highlight overview
This pull request primarily includes cosmetic changes, such as renaming methods and reorganizing packages. However, it also features significant refactoring of the `quarkus-spring-data-rest-extension` to align with the latest `spring-data-jpa 3.x` updates.

#### Key Changes in spring data jpa latest version:

**Introduction of New Interfaces:**
   - With `spring-data-jpa 3.x`, two new interfaces, `ListCrudRepository` and `ListPagingAndSortingRepository`, were introduced. These interfaces return `List<T>` types, unlike the existing `PagingAndSortingRepository` and `CrudRepository`, which return `Iterable<T>`.
   - The class hierarchy has been modified as follows: 

Captura de pantalla 2024-06-13 a las 14 42 32

- `PagingAndSortingRepository` **no** longer extends `CrudRepository`.
- `ListCrudRepository` extends `CrudRepository`. 
- `ListPagingAndSortingRepository` extends `PagingAndSortingRepository`. 
- `JpaRepository` extends both `ListCrudRepository`and `ListPagingAndSortingRepository`.

#### Refactoring of `spring-data-rest` Extension:
   - To accommodate these changes, the extension's hierarchy was restructured. Rather than maintaining different implementors, a single implementor now checks the repository type and implements the corresponding methods.
   - All logic is consolidated in the `RepositoryMethodsImplementor`, which implements all repository methods based on the repository type.
   - The `SpringDataRestProcessor` has been updated to include the new interfaces and now performs registration in a unified method (`registerRepositories`) instead of using two separate methods.
   - Unification of PropertiesProvider Classes: these classes have been unified to incorporate the new methods from the `List***Repository` interfaces.
   - Updates to `EntityClassHelper`: the `EntityClassHelper` has been relocated and now includes new methods used by `RepositoryMethodsImplementor` to detect the repository type for implementation.
   - Modifications to Tests: the `Paged***Test` no longer checks methods inherited from `Crud**Repos`. A new test case, `CrudAndPagedResourceTest`, has been added for a repository extending both `Crud` and `Paged` repositories.

Note:
The new `List****` methods are not specifically exposed via REST, as this is consistent with Spring's behavior.

Additionally, I plan to clean up some comments in the code. I will address these and any other changes based on your feedback during the review.

### CODE REVIEW

**Comment by aureamunoz** [comment]
cc @geoand

**Comment by geoand** [comment]
Thanks for this.

The CI failures definitely seem related to the PR

**Comment by aureamunoz** [comment]
> Thanks for this.
> 
> The CI failures definitely seem related to the PR

Oh, absolutely. I will take a look

**Comment by geoand** [comment]
Sure, I'll try take and take a look tomorrow

**Comment by geoand** [comment]
A high level overview of the changes in the description of the PR would be nice for future reference.

File: integration-tests/spring-data-jpa/src/main/resources/application.properties [hunk 1]
```diff
@@ -7,7 +7,7 @@ quarkus.hibernate-orm.database.generation=drop-and-create
 # DataGenerator uses java.util.Random => needs to be runtime initialized
 # as j.u.Random is prohibited in the native image heap. Otherwise
 # random data becomes deterministic
-quarkus.native.additional-build-args=--initialize-at-run-time=io.quarkus.it.spring.data.jpa.DataGenerator
+quarkus.native.additional-build-args=--initialize-at-run-time=io.quarkus.it.spring.data.jpa.DataGenerator\\,org.springframework.data.domain.Sort$TypedSort,--initialize-at-build-time=org.springframework.aop,-H:TraceClassInitialization=true,-H:AbortOnTypeReachable=org.springframework.data.util.MethodInvocationRecorder
```
[the remark this answers is not in the document]
Comment (aureamunoz) [reply integration-tests/spring-data-jpa/src/main/resources/application.properties:10]: OUps, yes, this is a leftover of when I was trying to fix the native problem. I will remove it

**Comment by geoand** [comment]
> A high level overview of the changes in the description of the PR would be nice for future reference.

I would also like to know why there are new implementors. Please add this information to the description of the PR

**Comment by aureamunoz** [comment]
> > A high level overview of the changes in the description of the PR would be nice for future reference.
> 
> I would also like to know why there are new implementors. Please add this information to the description of the PR

Yes, sure. It's in progress.

**Comment by aureamunoz** [comment]
### Highlight overview
This pull request primarily includes cosmetic changes, such as renaming methods and reorganizing packages. However, it also features significant refactoring of the `quarkus-spring-data-rest-extension` to align with the latest `spring-data-jpa 3.x` updates.

#### Key Changes in spring data jpa latest version:

**Introduction of New Interfaces:**
   - With `spring-data-jpa 3.x`, two new interfaces, `ListCrudRepository` and `ListPagingAndSortingRepository`, were introduced. These interfaces return `List<T>` types, unlike the existing `PagingAndSortingRepository` and `CrudRepository`, which return `Iterable<T>`.
   - The class hierarchy has been modified: `PagingAndSortingRepository` no longer extends `CrudRepository`. Instead, `JpaRepository` now extends both `CrudRepository` and `PagingAndSortingRepository`.

#### Refactoring of `spring-data-rest` Extension:
   - To accommodate these changes, the extension's hierarchy was restructured. Rather than maintaining different implementors, a single implementor now checks the repository type and implements the corresponding methods.
   - All logic is consolidated in the `RepositoryMethodsImplementor`, which implements all repository methods based on the repository type.
   - The `SpringDataRestProcessor` has been updated to include the new interfaces and now performs registration in a unified method (`registerRepositories`) instead of using two separate methods.
   - Unification of PropertiesProvider Classes: these classes have been unified to incorporate the new methods from the `List***Repository` interfaces.
   - Updates to `EntityClassHelper`: the `EntityClassHelper` has been relocated and now includes new methods used by `RepositoryMethodsImplementor` to detect the repository type for implementation.
   - Modifications to Tests: the `Paged***Test` no longer checks methods inherited from `Crud**Repos`. A new test case, `CrudAndPagedResourceTest`, has been added for a repository extending both `Crud` and `Paged` repositories.

Note:
The new `List****` methods are not specifically exposed via REST, as this is consistent with Spring's behavior.

**Comment by geoand** [comment]
Thanks, I updated the sescription of the PR

**Comment by geoand** [comment]
It is possible that the relocations / moves can be made so that git is aware of them and they don't appear as new files?

**Comment by aureamunoz** [comment]
> It is possible that the relocations / moves can be made so that git is aware of them and they don't appear as new files?

The only file marked as new in git that has been moved is Entity.java but besides being moved some things inside have changed (method visibility modifiers) and some methods added so it was not a simple move and git considers it to be a new class.

Then, `RepositoryMethodsImplementor`and `RepositoryPropertiesProvider`are new, they contain the logic that was previously dispatched between Crud and Paged implementors and providers

DefaultPagedResourceBisTest is a leftover and will be removed

**Comment by gsmet** [comment]
Yeah if the files are small and have too many changes, Git won't be successful in identifying it as a move.

**Review by geoand — APPROVED** [review]
Thanks!

**Comment by michalvavrik** [comment]
@geoand I added breaking-change, not sure why you removed it but this is super breaking. for example it breaks compilation due to _The class hierarchy has been modified: PagingAndSortingRepository no longer extends CrudRepository_ as written in description. it already broke QE project and if compilation is broken, it must be documented. Please add note to the migration guide.

**Comment by gsmet** [comment]
Thanks for your vigilance BTW!

**Comment by michalvavrik** [comment]
thank you @gsmet . BTW I have 4 failures after this PR got merged. I didn't use Spring for 3 years, so my knowledge is limited, so it could be false alarm. Anyway, it feels like a bug so I'll open issues within one hours.

**Comment by gsmet** [comment]
@michalvavrik you can create them tomorrow. We still have a week before the Core Final.

**Comment by geoand** [comment]
Thanks.

I removed it by mistake

**Comment by jedla97** [comment]
> The class hierarchy has been modified: `PagingAndSortingRepository` no longer extends `CrudRepository`. Instead, `JpaRepository` now extends both `CrudRepository` and `PagingAndSortingRepository`.

Reading this I'm not sure if it's correct. `JpaRepository` is now extended by `ListCrudRepository` and `ListPagingAndSortingRepository`. Yes `ListCrudRepository` is extended by `CrudRepository` but `JpaRepository` is not directly extend `CrudRepository`. 

Not sure if this should be updated in migration guide.

**Comment by aureamunoz** [comment]
> > The class hierarchy has been modified: `PagingAndSortingRepository` no longer extends `CrudRepository`. Instead, `JpaRepository` now extends both `CrudRepository` and `PagingAndSortingRepository`.
> 
> Reading this I'm not sure if it's correct. `JpaRepository` is now extended by `ListCrudRepository` and `ListPagingAndSortingRepository`. Yes `ListCrudRepository` is extended by `CrudRepository` but `JpaRepository` is not directly extend `CrudRepository`.
> 
> Not sure if this should be updated in migration guide.

It's  the opposite, I improved the description [here](https://github.com/quarkusio/quarkus/pull/40344#issue-2268984198). Take a look and let me know if it's clearer.

**Comment by aureamunoz** [comment]
A new PR is coming fixing the #41134 , #41135 and #41136 but it needs a quarkus-spring-data-api release to be done ([PR#14](https://github.com/quarkusio/quarkus-spring-data-api/pull/14))

**Comment by michalvavrik** [comment]
Cool, thanks!

**Comment by jedla97** [comment]
@aureamunoz Thanks for working on this and thanks for update. 

It's much more clearer. I know what's happend after reading it and don't need to deep dive how it changed.

Probably one thing is it should be also updated in https://github.com/quarkusio/quarkus/wiki/Migration-Guide-3.12

**Pull request merged**